add dp60 indicator mode

yulei opened the pull request:
<!--- Provide a general summary of your changes in the title above. -->
Add support for an open source viper/eagle FR4 plate(8 RGB indicator leds).
https://github.com/yulei/viper_plate
<!--- This template is entirely optional and can be removed, but is here to help both you and us. -->
<!--- Anything on lines wrapped in comments like these will not show up in the final text. -->

## Description

<!--- Describe your changes in detail here. -->

## Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->
- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ x] Keyboard (addition or update)
- [ x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x ] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [ x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

drashna commented on keyboards/dp60/keymaps/indicator/indicator.c:
```diff
@@ -0,0 +1,224 @@
+/*
```
I'm not sure that this file is needed, especially with the newer "lighting layers" code.

https://docs.qmk.fm/#/feature_rgblight?id=lighting-layers

yulei replied:
Yes, seems it can be used with the "lighting layers" codes, but it requires to customize the rgblight_set() function because a different pin was used for the leds on the plate.

yulei replied:
Currently the RGBLIGHT_LAYERS can not use with RGBLIGHT_CUSTOM_DRIVER due to a link error.
 ```
Compiling: quantum/rgblight.c                                                                      quantum/rgblight.c:638:13: error: 'rgblight_layers_write' defined but not used [-Werror=unused-function]
 static void rgblight_layers_write(void) {
             ^
cc1.exe: all warnings being treated as errors
```
It was caused by the current implementation of the light layers.
```
#ifndef RGBLIGHT_CUSTOM_DRIVER
void rgblight_set(void) {
    LED_TYPE *start_led;
    uint16_t  num_leds = clipping_num_leds;

#    ifdef RGBLIGHT_LAYERS
    if (rgblight_layers != NULL) {
        rgblight_layers_write();
    }
#    endif
```
riglight_layers_write() was a static function and can't be called out of rgblight.c

drashna replied:
Ah. 

No need to use custom driver anymore.  you can use `rgblight_call_driver` and handle it there. 

You can take a look at how the ergodox ez does this in the led_i2c.c file, in fact.

zvecr commented on keyboards/dp60/keymaps/indicator/indicator.c:
```diff
@@ -0,0 +1,253 @@
+#define w_nop4 w_nop2 w_nop2
+#define w_nop8 w_nop4 w_nop4
+#define w_nop16 w_nop8 w_nop8
+void inline indicator_sendarray_mask_ind(uint8_t *data, uint16_t datlen, uint8_t maskhi) {
```
Not a fan of this duplication of core code. Im guessing the only requirement is to be able to use 2 different RGB pins, and if so, theres ways to avoid having to re-implement the ws2812 driver.

yulei replied:
The current avr ws2812 driver embedded the PIN definition on the assembly codes. I'm not familiar with assembly and have no idea how to extract the PIN definition from the codes. Any help is great.

drashna replied:
IIRC, we've had this talk on discord (not you specifically), but there is no way to do this without ... basically re-implementing it, as the data pin HAS to be defined at compile time, and can't be changed in runtime. 

This is also why #9412 was opened.

drashna replied:
I think was the main thing holding this back. 

Another way to do this, maybe would be to undefine the rgbled pin, define the new version, and then include ws2812.c.  Which, is super hacky, but wouldn't duplicate core code.

drashna reviewed (changes requested):
Could you add a GPL2+ compatible license header to all of the `c` and `h` files? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

yulei commented:
> Could you add a GPL2+ compatible license header to all of the `c` and `h` files?
> 
> For instance:
> https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

Added license header for those files and moved the ws2812 driver code to a separate file. To avoid the link time error(duplicate function name), I have to rename the origin ws2812* to indicator*, thus if someone rename the function names on the ws2812.c. Those codes will be broken :(

drashna commented:
Looks like there is a merge conflict here, too. If you can resolve that?

noroadsleft commented:
Thanks!
